File: thread_pool.py
"""
Thread Pool Example
Demonstrates using ThreadPoolExecutor for efficient thread management.
"""
from concurrent.futures import ThreadPoolExecutor, as_completed
import time
import random


def task(task_id):
    """
    Simulated task that takes variable time to complete.
    
    Args:
        task_id: Unique identifier for the task
        
    Returns:
        Tuple of (task_id, result, duration)
    """
    duration = random.uniform(0.5, 2.0)
    print(f"Task {task_id} starting (will take {duration:.2f}s)")
    time.sleep(duration)
    result = task_id * task_id
    print(f"Task {task_id} completed")
    return task_id, result, duration


def run_thread_pool_example():
    """Run thread pool example with multiple tasks."""
    print("=== Thread Pool Example ===\n")
    
    num_tasks = 10
    max_workers = 3
    
    print(f"Executing {num_tasks} tasks with {max_workers} workers\n")
    
    start_time = time.time()
    
    # Using ThreadPoolExecutor
    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        # Submit all tasks
        futures = [executor.submit(task, i) for i in range(num_tasks)]
        
        # Process completed tasks as they finish
        results = []
        for future in as_completed(futures):
            task_id, result, duration = future.result()
            results.append((task_id, result))
            print(f"Retrieved result: Task {task_id} = {result}")
    
    end_time = time.time()
    total_time = end_time - start_time
    
    print(f"\nAll tasks completed in {total_time:.2f} seconds")
    print(f"Results: {sorted(results)}\n")


def map_example():
    """Demonstrate ThreadPoolExecutor.map() for batch processing."""
    print("=== Thread Pool Map Example ===\n")
    
    def square(n):
        """Simple function to square a number."""
        time.sleep(0.2)
        return n * n
    
    numbers = list(range(10))
    
    with ThreadPoolExecutor(max_workers=4) as executor:
        results = list(executor.map(square, numbers))
    
    print(f"Input:  {numbers}")
    print(f"Output: {results}\n")


if __name__ == "__main__":
    run_thread_pool_example()
    map_example()

File: src/main/java/com/multithreading/ThreadPool.java
package com.multithreading;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.concurrent.*;

/**
 * Thread Pool Example
 * Demonstrates using ExecutorService for efficient thread management.
 */
public class ThreadPool {

    static class Task implements Callable<TaskResult> {
        private final int taskId;

        public Task(int taskId) {
            this.taskId = taskId;
        }

        @Override
        public TaskResult call() throws Exception {
            Random random = new Random();
            double duration = 0.5 + random.nextDouble() * 1.5;
            long durationMs = (long) (duration * 1000);

            System.out.println("Task " + taskId + " starting (will take " + 
                             String.format("%.2f", duration) + "s)");
            Thread.sleep(durationMs);
            
            int result = taskId * taskId;
            System.out.println("Task " + taskId + " completed");
            
            return new TaskResult(taskId, result, durationMs);
        }
    }

    static class TaskResult {
        final int taskId;
        final int result;
        final long durationMs;

        public TaskResult(int taskId, int result, long durationMs) {
            this.taskId = taskId;
            this.result = result;
            this.durationMs = durationMs;
        }

        @Override
        public String toString() {
            return "Task " + taskId + " = " + result;
        }
    }

    public static void runExample() throws InterruptedException, ExecutionException {
        System.out.println("=== Thread Pool Example ===\n");

        int numTasks = 10;
        int maxWorkers = 3;

        System.out.println("Executing " + numTasks + " tasks with " + maxWorkers + " workers\n");

        ExecutorService executor = Executors.newFixedThreadPool(maxWorkers);
        List<Future<TaskResult>> futures = new ArrayList<>();

        long startTime = System.currentTimeMillis();

        // Submit all tasks
        for (int i = 0; i < numTasks; i++) {
            Future<TaskResult> future = executor.submit(new Task(i));
            futures.add(future);
        }

        // Retrieve results as they complete
        List<TaskResult> results = new ArrayList<>();
        for (Future<TaskResult> future : futures) {
            TaskResult result = future.get();
            results.add(result);
            System.out.println("Retrieved result: " + result);
        }

        executor.shutdown();
        executor.awaitTermination(1, TimeUnit.MINUTES);

        long endTime = System.currentTimeMillis();
        double totalTime = (endTime - startTime) / 1000.0;

        System.out.println("\nAll tasks completed in " + String.format("%.2f", totalTime) + " seconds");
        System.out.print("Results: [");
        for (int i = 0; i < results.size(); i++) {
            System.out.print("(" + results.get(i).taskId + ", " + results.get(i).result + ")");
            if (i < results.size() - 1) System.out.print(", ");
        }
        System.out.println("]\n");
    }

    public static void runFixedThreadPoolExample() throws InterruptedException {
        System.out.println("=== Fixed Thread Pool Map Example ===\n");

        ExecutorService executor = Executors.newFixedThreadPool(4);
        List<Integer> numbers = List.of(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);

        System.out.println("Input:  " + numbers);

        List<Future<Integer>> futures = new ArrayList<>();
        for (Integer num : numbers) {
            futures.add(executor.submit(() -> {
                Thread.sleep(200);
                return num * num;
            }));
        }

        List<Integer> results = new ArrayList<>();
        for (Future<Integer> future : futures) {
            try {
                results.add(future.get());
            } catch (ExecutionException e) {
                System.err.println("Task failed: " + e.getMessage());
            }
        }

        executor.shutdown();
        executor.awaitTermination(1, TimeUnit.MINUTES);

        System.out.println("Output: " + results + "\n");
    }

    public static void main(String[] args) throws InterruptedException, ExecutionException {
        runExample();
        runFixedThreadPoolExample();
    }
}
